internal/session/sessionwantsender.go

```diff
@@ -330,24 +349,23 @@ func (sws *sessionWantSender) processUpdates(updates []update) []cid.Cid {
 				sws.peerConsecutiveDontHaves[upd.from]++
 			}
 		}
+	}
 
-		// For each HAVE
+	// Process received HAVEs
+	for _, upd := range updates {
 		for _, c := range upd.haves {
+			// If we already received a block for the want, ignore any HAVE for
+			// the want
+			if blkCids.Has(c) {
+				continue
```

Comment:
Would this prevent us from adding this peer to the session?

Reply:
Adding the peer to the session happens before `processUpdates()` gets called. However this does interfere with consecutive DONT_HAVE counting so I refactored to avoid interference.